src/test/java/org/scijava/plugin/SingletonServiceTest.java

```diff
@@ -0,0 +1,131 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package org.scijava.plugin;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.scijava.Context;
+import org.scijava.convert.AbstractConverter;
+import org.scijava.convert.ConvertService;
+import org.scijava.convert.Converter;
+import org.scijava.plugin.event.PluginsAddedEvent;
+import org.scijava.plugin.event.PluginsRemovedEvent;
+
+/**
+ * Tests {@link AbstractSingletonService}.
+ * 
+ * @author Stefan Helfrich (University of Konstanz)
+ */
+public class SingletonServiceTest {
+
+	private PluginService pluginService;
+	private ConvertService convertService;
+
+	@Before
+	public void setUp() {
+		final Context context = new Context(PluginService.class,
+			ConvertService.class);
+		pluginService = context.service(PluginService.class);
+		convertService = context.service(ConvertService.class);
+	}
+
+	@After
+	public void tearDown() {
+		pluginService.context().dispose();
+	}
+
+	/**
+	 * Tests that the {@link AbstractSingletonService} properly handles
+	 * {@link PluginsAddedEvent}s originating from the {@link PluginService}.
+	 */
+	@Test
+	public void testSingletonServicePluginsAddedHandling() {
+		@SuppressWarnings("rawtypes")
+		PluginInfo<Converter> converterInfo = new PluginInfo<>(
+			DummyStringConverter.class, Converter.class);
+		converterInfo.setPluginClass(DummyStringConverter.class);
+
+		pluginService.addPlugin(converterInfo);
+
+		assertNotNull(pluginService.getPlugin(DummyStringConverter.class));
+		assertTrue(convertService.supports("StringInput", Number.class));
+	}
+
+	/**
+	 * Tests that the {@link AbstractSingletonService} properly handles
+	 * {@link PluginsRemovedEvent}s originating from the {@link PluginService}.
+	 */
+	@Test
+	public void testSingletonServicePluginsRemovedHandling() {
+		@SuppressWarnings("rawtypes")
+		PluginInfo<Converter> converterInfo = new PluginInfo<>(
+			DummyStringConverter.class, Converter.class);
+		converterInfo.setPluginClass(DummyStringConverter.class);
+
+		pluginService.addPlugin(converterInfo);
+
+		// De-register DummyStringConverter
+		pluginService.removePlugin(converterInfo);
```

Comment:
Instead of adding then immediately removing, a cleaner approach might be to simply remove one of the built-in discovered plugins. If you don't like that external dependency for some reason (and I wouldn't blame you), you could create a second static inner class here annotated with `@Plugin` so that it is automatically discovered and added initially, and then remove it here. The advantage is that we ensure it works to remove plugins where were present _when the context was first initialized_. I think this is a more robust test—or at least a needed additional test.